### PR TITLE
feat(controller): routes, handling function backbone and errors management 

### PR DESCRIPTION
**Pull request by thomas-mauran**
### routes, handling function backbone and errors management 

The goal here is to provide the backbone of the rest API to then only have to implement it. 
There are some TODOs in the code to split the work efficiently with my team.  

### Featuring
- error handling using `thiserror` and `anyhow`
- the following routes and their handling function definition

#### Routes

**GET** workloads/
**POST** workloads/
**GET** workloads/:id
**DELETE** workloads/:id

**GET** instances/
**POST** instances/
**GET** instances/:id
**DELETE** instances/:id

### CODE REVIEW

**Comment by thomas-mauran** [comment]
ping @sameo (sorry we can't add you as reviewer)

File: controller/src/main.rs [hunk 1]
```diff
@@ -1,25 +1,31 @@
...
 use tokio::task;
 
+use axum::routing::{delete, post};
+use log::info;
+use routes::instances::{delete_instance, get_instances, get_specific_instance, post_instance};
+use routes::workloads::{delete_workload, get_specific_workload, get_workloads, post_workload};
+
 #[derive(Debug, Default)]
-pub struct SchedulerService {}
+pub struct MySchedulingService {}
```
Comment (sameo) [inline controller/src/main.rs]: I much prefer `SchedulerService`than `MySchedulingService` tbh. Even simply `Scheduler`would be fine.

File: controller/src/types/workload_request.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#[derive(Debug, Deserialize, Serialize)]
+pub enum WorkloadKind {
+    Container,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum WorkloadRegistry {
+    Docker,
+    Podman,
+    Ghcr,
```
Comment (sameo) [inline controller/src/types/workload_request.rs:19]: What is Ghcr?

File: controller/src/main.rs [hunk 3]
```diff
@@ -35,25 +41,19 @@ impl SchedulingService for SchedulerService {
...
-                    status_code: 0,
-                    message: "Workload 1 is terminated".to_string(),
-                    ..Default::default()
-                },
-                WorkloadStatus {
-                    name: "Workload 2".to_string(),
-                    status_code: 0,
-                    message: "Workload 2 is running".to_string(),
+                    status_code: 1,
+                    message: "Your workload is RUNNING".to_string(),
```
Comment (sameo) [inline controller/src/main.rs:49]: FWIW, it's not really useful to pass a string message back as the status. The status should have all necessary information for the caller/receiver to build any logging message it wants to display. Something to be fixed later.

File: controller/src/main.rs [hunk 4]
```diff
@@ -63,6 +63,9 @@ impl SchedulingService for SchedulerService {
                     .send(Ok(status))
                     .await
                     .expect("Failed to send status to stream");
+
+                // Attendre 10 secondes avant le prochain envoi
+                sleep(Duration::from_secs(10)).await;
```
Comment (sameo) [inline controller/src/main.rs]: Why? And also, please don't use french comments if the rest of the code is using english.

File: controller/src/main.rs [hunk 5]
```diff
@@ -81,59 +84,48 @@ impl SchedulingService for SchedulerService {
 
 #[tokio::main]
 async fn main() -> Result<(), Box<dyn std::error::Error>> {
-    // // Initialize logger
+    // SETUP LOGGING
```
Comment (sameo) [inline controller/src/main.rs]: Why are we changing perfectly fine comments to new, upper case ones?

File: controller/src/main.rs [hunk 6]
```diff
@@ -81,59 +84,48 @@ impl SchedulingService for SchedulerService {
...
+        )
+        .route("/instances", post(post_instance).get(get_instances))
+        .route(
+            "/instances/:id",
+            delete(delete_instance).get(get_specific_instance),
+        );
 
     // Spawn the HTTP server as a tokio task
     let http_thread = task::spawn(async move {
+        info!("Running http here: {}", http_addr);
```
Comment (sameo) [inline controller/src/main.rs]:
```suggestion
        log!("HTTP server running at {}", http_addr);
```

File: controller/src/routes/workloads.rs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    // Create a grpc workload object
+    let workload = Workload {
+        name: json_body.workload.name,
+        r#type: Type::Container.into(),
+        image: json_body.workload.image,
+        environment,
+        resource_limits: Some(Resources {
+            cpu: Some(1_i32),
+            memory: Some(1_i32),
+            disk: Some(1_i32),
```
Comment (sameo) [inline controller/src/routes/workloads.rs]: Where are those default values coming from? If you want to build a default `Resources`instance, it should implement the `Default`trait.

**Review by sameo — COMMENTED** [review]
Some nits. Overall it looks good, but please:

1. Squash your cleanup/fixes/linting commits (e.g. https://github.com/dev-sys-do/orka/pull/16/commits/7568cfe690a92120acd528e449be7f09760f6f76)
2. Add some actual commit message (not just a subject) to your commits. They are doing some meaningful changes, and the commit messages should describe them.

File: controller/src/types/workload_request.rs (hunk 2 above)
Comment (thomas-mauran) [reply controller/src/types/workload_request.rs:19]: Ghcr is the github registry it is very useful to keep an image linked to a repo

File: controller/src/routes/workloads.rs (hunk 7 above)
Comment (thomas-mauran) [reply controller/src/routes/workloads.rs]: Done

File: controller/src/main.rs (hunk 6 above)
Comment (thomas-mauran) [reply controller/src/main.rs]: done

File: controller/src/main.rs (hunk 5 above)
Comment (thomas-mauran) [reply controller/src/main.rs]: done

File: controller/src/main.rs (hunk 4 above)
Comment (thomas-mauran) [reply controller/src/main.rs]: done removed since it was not useful

File: controller/src/main.rs (hunk 3 above)
Comment (thomas-mauran) [reply controller/src/main.rs:49]: this message is based on the protobug from the scheduler team we will have to check with them to remove it there

File: controller/src/main.rs (hunk 1 above)
Comment (thomas-mauran) [reply controller/src/main.rs]: done

**Comment by thomas-mauran** [comment]
> Some nits. Overall it looks good, but please:
> 
>     1. Squash your cleanup/fixes/linting commits (e.g. [7568cfe](https://github.com/dev-sys-do/orka/commit/7568cfe690a92120acd528e449be7f09760f6f76))
> 
>     2. Add some actual commit message (not just a subject) to your commits. They are doing some meaningful changes, and the commit messages should describe them.

done

**Review by sameo — APPROVED** [review]
Please rebase your PR on top of the main one and we'll merge it.

**Comment by thomas-mauran** [comment]
> Please rebase your PR on top of the main one and we'll merge it.

done let me know if i did the right thing

**Comment by sameo** [comment]
> > Please rebase your PR on top of the main one and we'll merge it.
> 
> done let me know if i did the right thing

You did, thanks.

**Pull request merged**